include/circt/Analysis/OpCountAnalysis.h

```diff
@@ -0,0 +1,43 @@
+//===- OpCountAnalysis.h - operation count analyses -----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This header file defines prototypes for methods that perform analysis
+// involving the frequency of different kinds of operations found in a
+// builtin.module.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_ANALYSIS_OPCOUNT_ANALYSIS_H
+#define CIRCT_ANALYSIS_OPCOUNT_ANALYSIS_H
+
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/Operation.h"
+#include "llvm/ADT/DenseMap.h"
+
+namespace mlir {
+class AnalysisManager;
+} // namespace mlir
+namespace circt {
+namespace analysis {
+
+struct OpCountAnalysis {
+  OpCountAnalysis(Operation *moduleOp, mlir::AnalysisManager &am);
+
+  size_t getOpCount(OperationName opName);
+  SmallVector<OperationName> getFoundOpNames();
+  DenseMap<size_t, size_t> getOperandCountMap(OperationName opName);
+
+private:
+  DenseMap<OperationName, size_t> opCounts;
+  DenseMap<OperationName, DenseMap<size_t, size_t>> operandCounts;
```

Comment:
`opCounts` could be calculated from `operandCounts` here, but it would require a lot of extra iterating over `DenseMap`s so went for the faster option since we probably don't want this to be slower than necessary if it's running on huge blocks of IR.